### DISCUSSION TITLE
Improve the shortcuts.json format

### DISCUSSION THREAD

**Discussion by friday** [Ulauncher/Ulauncher#880] [Ideas]
The shortcuts data is placed in `{CONFIG_DIR}/shortcuts.json` (most likely `~/.config/ulauncher/shortcuts.json`) and will look something like this:

```json
{
    "4e47a6c6-1ff3-4e9c-a759-83a30af14d6c": {
        "id": "4e47a6c6-1ff3-4e9c-a759-83a30af14d6c",
        "name": "Google Search",
        "keyword": "g",
        "cmd": "https://google.com/search?q=%s",
        "icon": "/usr/share/ulauncher/media/google-search-icon.png",
        "is_default_search": true,
        "run_without_argument": false,
        "added": 1636503936.6266034
    },
    "8fb5ec80-d2ab-48fb-a7f1-6a0dca4d7a15": {
        "id": "8fb5ec80-d2ab-48fb-a7f1-6a0dca4d7a15",
        "name": "Stack Overflow",
        "keyword": "so",
        "cmd": "https://stackoverflow.com/search?q=%s",
        "icon": "/usr/share/ulauncher/media/stackoverflow-icon.png",
        "is_default_search": true,
        "run_without_argument": false,
        "added": 1636503936.6266315
    },
    "0e5a507a-409c-4b46-8d6f-b7b9ba62b78e": {
        "id": "0e5a507a-409c-4b46-8d6f-b7b9ba62b78e",
        "name": "Wikipedia",
        "keyword": "wiki",
        "cmd": "https://en.wikipedia.org/wiki/%s",
        "icon": "/usr/share/ulauncher/media/wikipedia-icon.png",
        "is_default_search": true,
        "run_without_argument": false,
        "added": 1636503936.6266506
    }
}
```

I see several problems with this:
* It's complicated to read and edit, should anyone want to do so from outside of Ulauncher.
* Why do we store "added"? It's not something we actually use.
* We have both names and keywords which has to be unique. So why do we add these really complicated UUIDs on top of that (and store them twice)? This also makes it harder to share them, and it's not needed imo. If a user wants two shortcuts for YouTube I think it's fair to assume only one will be the main YouTube shortcut and the other will be some specialized thing like "YouTube Subscriptions Search" with a different name and shortcut.
* "run_without_argument" is a useless property. We can just check if cmd contains %s. It's easier for everyone and more consistent, but I was voted out in this in #191

As I see it, this would be much better:

```ini
[g]
Name=Google Search
Exec=https://google.com/search?q=%s
Icon=/usr/share/ulauncher/media/google-search-icon.png
ShowInSearch=true

[so]
Name=Stack Overflow
Exec=https://stackoverflow.com/search?q=%s
Icon=/usr/share/ulauncher/media/stackoverflow-icon.png
ShowInSearch=true

[wiki]
Name=Wikipedia
Exec=https://en.wikipedia.org/wiki/%s
Icon=/usr/share/ulauncher/media/wikipedia-icon.png
ShowInSearch=false
```

This is the INI format. INI is a real config format allowing comments and custom formatting (unlike JSON). Python has a [built in INI parser/writer](https://docs.python.org/3/library/configparser.html) that can read and write this easily.

Part of me would like to store these as separate files instead, to make them easy to share between users and installs, but then we would get tiny files and need a useless generic section name (the part between the square brackets)

**Comment by troycurtisjr** [comment]
I suppose the right answer depends on what we want to treat this file as. Is this a serialized form of our configuration, which is driven by a GUI configuration dialog, or is this a (text based, configuration oriented) user interface. If it is just a serialized configuration, then json works nicely because it trivially fits with the built in data structures in Python. As an interface that we expect the user to use, it certainly leaves a lot to be desired.

As a user interface for sure the `added` field isn't directly used and would be a pain to add and the UUID is similarly difficult to work with as a user. `added` is just a useful bit of metadata which is trivially retained should we want to reference it for some reason. I know I've learned to always include created/last-updated columns in my database schema, this seems along the same idea, but certainly doesn't seem strictly required. 

The nice part of the UUID is that is allows the abstract concept of the particular configuration item to be distinct from the specific properties. Instead of the "object" changing when say the "Name" changes. In our case, the name-keyword-url are likely extremely tightly coupled anyway, so there probably isn't a big advantage in maintaining the distinction. If a particular config name is changed, and name is the "key", then it isn't much of an issue to have a "different" object. A slight hassle if there is a dict keyed off of one of the mutable properties, since it requires a delete and re-add, but certainly manageable.

On the format front, I think the biggest question (aside from whether the intention is to have this as a user interface or not) is what additional configurations we foresee. For example, if we end up considering doing "shortcuts" which also allow running scripts with args instead of just URLs, the INI escaping/parsing might be a little more limiting. For instance, even if we didn't allow full scripts, but instead just a command line, it might be nice to have true array support, like we'd have with json (or yaml) to enable specifying in an "execv" type of format. The alternative would be to require "command line" syntax with shell escaping semantics, but the shell escaping is probably very similar to any escaping that the configuration file might have, leading to double escaping. 

Regarding comments and custom formatting. I think that only works easily if the config is read-only for ulauncher. Maintaining the format of all the "other" stuff while using the preferences window is likely going to range from a pain to impossible, especially when things are added from the UI. Assuming the parser retains the formatting and comments for us after we read it in, where are comments "attached"? If there are comments before a section header, is that a comment for the following (probably) or preceding (possibly) section? If we decide that it is for the following section, what does it mean if the user changes the keyword for that section? Is it a different thing now and we should purge the comments, or did they just tweak something and we need to keep the comments as is? For maximum friendliness for version control (certainly not strictly required), we should probably have a deterministic ordering of sections (assuming the ini code can do that?), which means we'd have to decide how to handle surrounding formatting if we insert a new section. Using "insertion" order might help a good deal with all of these cases, since we wouldn't need to breakup sections/comments on renames (the section just stays in the same place) or on inserts (we just add to the end). But we have the problem again when deleting a section. What comments should be deleted with it?

When I started this comment, I wasn't for or opposed to the change or for leaving it, I just wanted to talk through the implications and use cases. However, now that I've typed out the issues around comments/formatting, I'm not sure how we reasonably support arbitrary user data while still supporting the ability of writing out the config from the dialog, at least not without a lot of assumptions that are likely to be untrue in many cases.  So while I don't have a strong preference between using json or INI as file formats for this particular usage, it feels like INI has the potential to bring in a lot of issues and surprising behavior, at least if we continue to allow changes from the preferences UI. Dropping modifications from the preferences UI would definitely be something I would NOT be a fan of.

**Comment by friday** [comment]
> Dropping modifications from the preferences UI would definitely be something I would NOT be a fan of.

Definitely wasn't planning on that.

> On the format front, I think the biggest question (aside from whether the intention is to have this as a user interface or not) is what additional configurations we foresee.

Nothing imo. It's meant to be simple. If it gets more complicated it will start to mimic apps or extensions, and then it's better people use apps or extensions. I even think it's bit unclear now when/why to use one vs the other.

> For example, if we end up considering doing "shortcuts" which also allow running scripts with args instead of just URLs, the INI escaping/parsing might be a little more limiting.

Shortcuts already do that: https://github.com/Ulauncher/Ulauncher/blob/02255436819a32833c75025283e294dc71a06691/ulauncher/search/shortcuts/ShortcutResultItem.py#L78-L90

> For instance, even if we didn't allow full scripts, but instead just a command line, it might be nice to have true array support, like we'd have with json (or yaml)

I agree that it's possible to add user input to the UI that stores some kind of array data. For example a tag list for adding alternative ways to find the shortcut through search. Then we would need to add a Vue component that lets users edit tag lists and we would need to send the data from the Preferences app (Vue/Js) via jsonp.js that converts the data to URL params and sends to Ulauncher (Python) via the Router.py, which converts it back. Neither of them support nested structures.

Previously (or currently in the dev branch) there was one component that used array input. This was the directories input component we used for the blacklisted app dirs. This was handled by joining and splitting the array with a colon delimiter on the JS side that turned them into a PATH like string.
Maybe this issue could be refactored away by sending JSON instead of URL params though. I'm kind of tempted to do that anyway because we are losing type information now anyway, and that leads to some redundant type conversion code.

I also agree that the config parser is "lossy". It won't retain comments when you read and write to it, so those would indeed be lost unless the file is write protected by the users, or if we override the parser's `comment_prefixes` so comments are not treated as such (but this probably create more issues as you mentioned). So maybe having the ability to comment and indent is a con more than a pro, but I definitely don't think that's a showstopper either way.

It also won't retain indentation. If you try to store a multi line string with indentation, write it out to a file and read in the file again, the indentation simply won't be there. I doubt this will happen a lot, if ever, but it is a practical con with INI that JSON doesn't have.

INI/configparser does retain the order you define things in though, assuming you don't override that. JSON probably does not, but I haven't tried.

Anyway, what I disagree with is that the ability to store nested structures would affect the implementation or safety of running the scripts. The script / url property is one string (with an `%s` in it), represented by one textarea in the UI, and the argument is also one string. Ulauncher doesn't support the ability to pass multiple arguments to shortcuts. If it did, how would Ulauncher know how to separate the query into a list? That's something shortcuts must handle themselves if they need to, and they can now.

**Comment by troycurtisjr** [comment]
Wah? How did I miss that shortcuts can be scripts! Sure enough! Mostly that was exploration about what might be useful under json that we'd lose going to INI. It certainly isn't impossible to implement or anything, It is just that I've spent a disturbing amount of my shell-scripting career fighting "shell escaping" when passed from configs or upper level scripts down to other script :grin: I always use execv whenever possible! Though to be honest a UI to input such a argument array would be pretty unfriendly.

If the file format is not really intended for human manipulation, then it doesn't seem like there is much of a reason to switch formats. If we do switch, and even if we don't necessarily intend for users to generate their own tweaks with comments, they will. And then all their comments will be lost on the first time they touch the preference page. To me, such surprising behavior should probably be a show stopper unless there are really compelling reasons to switch. Well, I guess we could detect the presence of comments and abort any saves. That would be simple, prevent ulauncher from blowing away user changes, but still allow them to manage it themselves if they want.
